### ISSUE TITLE
[ENHANCEMENT]: Recommended way to install and run X11 GUI.

### ISSUE THREAD

**Issue by twaik** [termux/termux-packages#19197]
Hello. 
People who want to run X11 GUI in termux come to discord `termux/general` and matrix `termux/termux` very often. They are not Linux enthusiasts, but even if they are, they still have problems running GUI. 
The reason for this is pretty much simple: termux's way of running GUI is not very straightforward.
People do not really want to play with multiple X servers and DEs. They want to simply run `startx` or something simmilar and enjoy graphical environment. 
Or someone is not experienced enough and follows guides in Internet (which are malicious in some cases, unfortunately). 
So, IMHO, we should do the following things (maybe all of them, maybe not).

- Write a script OR make termux packages depend on each other to install and configure all GUI automatically. I.e. we can make tigervnc, termux-x11, Xfvb and other X servers (if available) implement virtual package `xserver` and make some of them installable by default, or maybe make a package `xserver` containing script which goes through available X servers and starts one of them. Also we should make X servers make an alternative to `$PREFIX/bin/X`. Or we can port real `startx` (I am not sure if it implements switching between available X servers). Also we should make all DEs make an alternative to `$PREFIX/bin/x-window-manager` and `$PREFIX/bin/x-session-manager` for easier switching between DEs.
- Write a script which configures GUI automatically. I mean script should install everything automatically or let user choose DE/WM/smth and X server with `dialog` (like `termux-change-repo` does). I am not sure if it can run during dpkg `post-install` because of apt lock (`$PREFIX/var/lib/dpkg/lock`?).
- We should merge `x11-repo` to main repo. Also there are enough packages requiring (both QT and regular) which are CLI only but depend on packages in `x11-repo` and can be used without X server (I do not remember names, but still relevant).

Also something recommended but not related to GUI directly.

- We should make `apt` ask about `apt update` in the case of interactive console (to avoid prompting in scripts) and only in the case if last update was more than 6 hours ago (because auto-updater launching has this period) and chain it to the start of package installing process. I suggest to patch the `apt` and not `pkg` intentionally. This will make people update environment more often. That is related to both `rolling release` model and to `partial upgrade` problem. I am not sure how exactly it is done in other `rolling release` distros.
- We should make automatic correcting releases for both github and f-droid to make it used up-to-date bootstrap and, again, avoid `partial upgrade` problem. Or maybe remove bootstrap from apk and force downloading it from Internet or sdcard (in the case if Internet is not available).
- We should add some newbie-friendly guides to our site for most popular cases, i.e. updating environment, installing packages, installing GUI, explaining basics, etc. I know we are not Linux kindergarten, but many people have ZERO knowledge about Linux and want to get started. It is not neccessary, but people like that can make A LOT of noise in discord.

I did not post it in Discussions intentionally. That really seems like our flaw.

One or all points mentioned here should make termux more newbie-friendly, but I am not someone who makes decisions here (at least not make decisions alone).

I am pinging the whole team and locking conversation only for collaborators. I really want to hear opinions of everybody to avoid misunderstanding.
@termux/app-maintainers @termux/package-maintainers

**Comment by finagolfin** [comment]
Sounds good, it is just up to someone who wants to do the work. I have never tried using the X support, so not going to be me.

**Comment by twaik** [comment]
> Do you really think that ~Linux kindergarten~ human behavior is fixable programmatically? It can't be fixed this way.

I agree, we can not fix human behaviour. But we can fix/avoid reasons they complain about (at least some of them).

> I agree that we can make a script for automating setup of desktop environments.

We can simply use `update-alternatives` mechanism and make some `dialog` GUI for this. No need to implement something new. Most DEs can set `x-session-manager` alternative which can be used by our startx. X server can be configured via `update-alternatives` too. Out script (in perfect case) should only invoke `apt update && apt install`, `update-alternatives` and `dialog`.

**Comment by twaik** [comment]
> * I don't know how this can look like, maybe packaged as custom bootstrap archives.

No need for custom bootstraps. We may simply install everything with our magic script.
About how it should look like: when you install Ubuntu or Debian most needed tools (WM, DE, file manager, text editor) are already set up and ready to be launched. I think we should have simmilar setup. Maybe we should not install everything, we may do something like flavors (like xfce-desktop, xfce-minimal, mate-desktop, mate-minimal). 



> Instead of patching `apt`, we can just hide it (remove from PATH). Users will have to use `pkg` then.

I agree. We can move it somewhere to `$PREFIX/libexec`.

**Comment by TomJo2000** [comment]
I think something along the lines of a display manager utility could work quite well for this.

I don't wanna start any bike-shedding, but let's say we provide something like a `termux/termux-sessions` repo which contains setup/startx scripts for various Desktops and Window Managers (as well as having the component packages as dependencies for each).
Say a `termux-xfce-session`, `termux-sway-session`, `termux-lxqt-session`, etc.

I guess a closer equivalent would be the "session files" used by Display Managers usually stored in `/usr/share/{wayland,x}-sessions` on regular distros.

This way users can install a desktop (or WM) of their choice and use a standardized utility to properly start any of their installed desktops.
Maintenance burden should be manageable past the initial effort of writing and testing sessions.
Though I recognize this shifts the burden setup verification mostly onto the session setup scripts.
Overall from a user perspective it should make the whole experience very modular, transparent and extensible;
over something like a "Spins" or "Flavors" approach which is more so the impression I am getting from the custom bootstraps concept.
Though the two are of course not mutually exclusive.

**Comment by TomJo2000** [comment]
I'd like to address the general packaging points one by one.

---

> We should merge x11-repo to main repo. Also there are enough packages requiring (both QT and regular) which are CLI only but depend on packages in x11-repo and can be used without X server (I do not remember names, but still relevant).

I disagree about scrapping the x11-repo entirely, it's a useful distinction.
Perhaps instead including the `x11-repo` [stub package](https://github.com/termux/termux-packages/blob/master/packages/x11-repo/build.sh) as part of the default apt repos shipped in the bootstrap, so users don't need to separately install it, makes more sense.

> - We should make `apt` ask about `apt update` in the case of interactive console (to avoid prompting in scripts) and only in the case if last update was more than 6 hours ago (because auto-updater launching has this period) and chain it to the start of package installing process. I suggest to patch the `apt` and not `pkg` intentionally. This will make people update environment more often. That is related to both `rolling release` model and to `partial upgrade` problem. I am not sure how exactly it is done in other `rolling release` distros.

I agree that we should make it clearer that partial updates are unsupported on Termux and perhaps a warning or patch to do a full-upgrade instead should be included, though I'm unsure how exactly this should be handled.
As for "reminding users" that updates are available, I'm not sure how I feel about that.
Perhaps we could bring back the dynamic MOTD concept from a while ago to remind users about updates, but we should never force or "nag" users to update. Though a full upgrade when installing new packages should obviously remain the recommendation, and a notice if attempting to install new packages with pending updates seems appropriate.

> - We should make automatic correcting releases for both github and f-droid to make it used up-to-date bootstrap and, again, avoid `partial upgrade` problem. Or maybe remove bootstrap from apk and force downloading it from Internet or sdcard (in the case if Internet is not available).

I am entirely unsure how you propose to manage, and for that matter distribute this. Perhaps I just need some more clarification. I'm not opposed to the idea in principle.

> - We should add some newbie-friendly guides to our site for most popular cases, i.e. updating environment, installing packages, installing GUI, explaining basics, etc. I know we are not Linux kindergarten, but many people have ZERO knowledge about Linux and want to get started. It is not neccessary, but people like that can make A LOT of noise in discord.

As far as I know a wiki revamp is already underway at termux/termux-community#1, and I fully agree that a update to our official resources is sorely needed.
Though that is obviously a time and labor intensive process.

**Comment by twaik** [comment]
> I am entirely unsure how you propose to manage, and for that matter distribute this. Perhaps I just need some more clarification. I'm not opposed to the idea in principle.

The main idea is to make `apk` users get from Github or F-Droid always be updated.

> I disagree about scrapping the x11-repo entirely, it's a useful distinction. Perhaps instead including the `x11-repo` [stub package](https://github.com/termux/termux-packages/blob/master/packages/x11-repo/build.sh) as part of the default apt repos shipped in the bootstrap, so users don't need to separately install it, makes more sense.

I agree.

**Comment by TomJo2000** [comment]
> The main idea is to make `apk` users get from Github or F-Droid always be updated.

That would work for F-Droid - *Provided they are pulling the releases*.
But for GitHub it would require essentially switching to weekly(?) releases for the main App, no?
I'm actually not 100% clear on how the bootstrapping process works in detail.
Now that I think about it, shouldn't it just pull down the latest bootstrap when first running the app?

**Comment by twaik** [comment]
> Now that I think about it, shouldn't it just pull down the latest bootstrap when first running the app?

AFAIK termux-app unpacks bootstrap contained in apk. It is contained inlined to [shared library](https://github.com/termux/termux-app/blob/master/app/src/main/cpp/termux-bootstrap-zip.S). 

Bootstrap is only applied when termux runs first time or it's data is reset.

> In any case I don't think we should force pre-existing users to install the latest bootstrap.
> Maybe a notification when a new bootstrap is released to remind them that updates are available?

Irrelevant to existing users. 

Newer bootstrap means downloading less updates while doing `pkg update` first time in fresh environment.

**Comment by agnostic-apollo** [comment]
My thoughts on this.

- Primary reason this is an issue is lack of good documentation provided by Termux itself which can be quickly linked when such a question is asked. There are lot of things that are reasonably documented, but we still get questions for them, that's never going to stop, but it's easier to handle them quickly without wasting time. Second reason is that documentation is so scattered, we have a wiki, termux site, each repo has it's own readme, and then some repos have "hidden" wikis as well, and some documentation is in random comments/issues, many by me. The termux site in future would be the central place for all the documentation and we can have guides for specific stuff that different users would be interested in, videos can also be linked in such guides to our youtube channel (where twaik also comes on in person with his face...). There should also be in app documentation, that's available locally to help users for specific things, internet is not always available for many people or is expensive. This is why I have spent tonne of time these last two years on fixing documentation design, it's mostly almost there as a framework, apart from the documentation itself of course, which also takes countless hours in itself to write accurately and detailed, and I obviously I can't do all of it alone, so other termux devs would have to step in when the time comes for their respective niches. With the site design, we would be able to resume community contributions too, so that may be helpful too, depending on their quality, wiki has had lot of contributions too.
- Modifying `apt` to auto `apt update` is not a good idea. We should stay consistent with other linux distros like debian and there are reasons `update` is not automatically done. It obviously wastes bandwidth, but it wastes time as well, I rarely use `pkg` myself and if I am installing multiple packages, it would be annoying to wait every time for sync. It also gives control to be able to install an old version of the package, in case newer is broken, or be able to reinstall a package from a local cache while testing. Such core behavior shouldn't be changed. For experienced devs who know what they are doing, `apt` should function as is. The solution instead is to be more vocal about using `pkg` in our docs for normal users, and recommend `apt` or `pacman` for advance users. I am very much against forcing updates on users.
- Neither `apt` or `pacman` should be removed from `$PATH`. Many people use it and it would break lot of user space programs and having to specially configure to make them accessible would be a chore for new installations.
- `apt` can however be patched to print a red warning that `pkg` should be used or `apt update` should be used first. But I doubt that will help for noobs, since that is currently already logged if package is not found, just a day ago we got a user on reddit, who took a snippet of the message from `apt` telling them to run `apt update` and they posted instead and asked what to do instead of just running the command as suggested. But there would be lot of people who actually read that message and ran the command and never posted about it, we never get to know about those, so it may be a biased view that such messages are not helpful to anyone.
- I agree that we can provide scripts that setup a DE for a user and allow them to switch. I agree with TomIO too that `x11-repo` could be added to bootstrap, lot of users do install termux looking for a DE, so it could be a good idea to provide the repo from the get go. The issue with that is that it would increase `apt update` time so may not be worth it if scripts automatically install it or docs already tell users to install it.
- Using bootstraps to provide an "experience" is not a good idea. What if user installs said bootstrap, and then installs unrelated packages, and then has to update the bootstrap, that would result in removal of his other packages and they would need to be installed and setup again, and for every update. Such upgrades should be handled by the scripts themselves, they can even run `apt upgrade` after confirming with user, this way packages wouldn't break.
- Termux app bootstrap being outdated is not that big an issue and not the cause of said problems. The one from 2 years ago still works fine on new installation, apart from `pkg` selecting broken mirrors sometimes and requiring update of `termux-tools` but that will be fixed with pre-release. Moreover, weekely releases just for bootstraps would be a release spam and it doesn't help with broken packages issue. Even if user installs latest bootstrap today, a few hours later, the repo may get a package that requires dependencies updates, so if user only installs that package without upgrading the dependencies, he would still get in that broken state. Also F-Droid takes days to update app, so any latest bootstrap packed would be stale by then anyways, depending on how many packages got upgraded before release and installation. The solution is to ask user if they want to run `apt upgrade` when bootstrap is installed, it will be added in a UI dialog when I work on bootstrap redesign, lot of other reqs exist too related to bootstrap, including custom ones. It will also support downloading the latest bootstrap if user wants too. However, both upgrade or latest bootstrap have issues too, such untested upgrades or bootstraps could result in broken states. It has happened in the past that broken bootstraps kept getting released for weeks and no one noticed. I can't test if a bootstrap is working every week, in the past, I manually thoroughly tested bootstraps before each app release to know for sure that they were stable. Automated testing would help solve some issues, `termux-exec-tests` being added should help too for testing basic execution environment.
- Bootstrap docs are available at https://github.com/termux/termux-packages/wiki/For-maintainers#bootstraps

**Comment by agnostic-apollo** [comment]
Btw, I plan on writing a guide for setting up a DE, `termux-x11` and firefox for mozilla people this week, finishing up some other work currently.

**Comment by agnostic-apollo** [comment]
>Newer bootstrap means downloading less updates while doing pkg update first time in fresh environment.

The bootstrap is like 25MB only for each architecture, downloading archives of outdated packages wouldn't take that long. And if we download latest bootstrap, same thing basically. One time setup time is irrelevant anyways.

**Comment by robertkirkman** [comment]
"partial upgrade" is mentioned repeatedly in this issue, so I will link this. If someone is interested in solutions to the "partial upgrade" problem and doesn't know about this yet, please consider testing and reviewing my idea and give feedback about how well it works for you if you want:

- https://github.com/termux/termux-tools/pull/172

> Command `install` should always check/refresh index and pull all system updates before installing the actual package.

@sylirre unfortunately your idea of patching `apt` itself is not ideal in my opinion and basically the reason why is that I think `apt` should continue to be available and behave as it currently does, which is the same as how `apt` behaves in the rolling-release distro Debian Sid and is what experienced users expect to work if they need it.

I believe the problem can be resolved by focusing on fixing the behavior of command `pkg install` in **`pkg` wrapper alone** to be significantly more helpful for beginner users than it currently is, particularly because the current behavior of the `pkg install` command is especially unfortunate and is itself (`pkg`, not `apt`) the primary cause of many issues users experience.

Also, with sufficiently careful changes to the design of the wrapper, it is possible to fix all bad behaviors without actually _always_ refreshing the package index. You can review my PR to understand the details of how that is possible.

> Modifying apt to auto apt update is not a good idea. We should stay consistent with other linux distros like debian and there are reasons update is not automatically done. It obviously wastes bandwidth, but it wastes time as well, I rarely use pkg myself and if I am installing multiple packages, it would be annoying to wait every time for sync. It also gives control to be able to install an old version of the package, in case newer is broken, or be able to reinstall a package from a local cache while testing. Such core behavior shouldn't be changed. For experienced devs who know what they are doing, apt should function as is. The solution instead is to be more vocal about using pkg in our docs for normal users, and recommend apt or pacman for advance users. I am very much against forcing updates on users.

basically I agree with that and my idea is also compatible with that opinion. Also I know that you can't review the PR right now, that's fine, this message is just for any others in this thread or reading who don't know about that PR yet.
